### PR TITLE
chore(xtest): Shared Scenario/Instance Pydantic schema in otdf-sdk-mgr

### PR DESCRIPTION
**Pull request by dmihalcik-virtru**
## Summary

First PR in a five-part stack that introduces a multi-instance test harness and a Claude plugin for OpenTDF bug reproduction. This PR adds *only* the shared Pydantic schema in `otdf-sdk-mgr` — no consumers yet.

- Adds `otdf_sdk_mgr.schema` with v2 models: `Scenario`, `Instance`, `PlatformPin`, `KasPin`, `SdkPin`, `ScenarioSdks`, `Suite`, etc.
- `ScenarioSdks.encrypt` / `.decrypt` mirror xtest's existing `--sdks-encrypt` / `--sdks-decrypt` convention so a→b-only scenarios are first-class.
- `python -m otdf_sdk_mgr.schema validate <path>` validates either a Scenario or an Instance file based on its `kind:`.
- Adds `pydantic` + `ruamel.yaml` to `otdf-sdk-mgr/pyproject.toml`.
- 6 unit tests covering round-trips, pin invariants, and unknown-field rejection.

## Stack

1. [**This PR**](https://github.com/opentdf/tests/pull/450) — Shared schema
2. [Platform installer + `install scenario`](https://github.com/opentdf/tests/pull/451) in `otdf-sdk-mgr` (builds on this)
3. `otdf-local` [multi-instance refactor](https://github.com/opentdf/tests/pull/452) + new CLI subcommands
4. `xtest/conftest.py` [integration](https://github.com/opentdf/tests/pull/453) (`--scenario`, `--instance`)
5. [Claude plugin](https://github.com/opentdf/tests/pull/454) (`.claude/skills/`, settings, plugin manifest)
6. https://github.com/opentdf/tests/pull/455

## Test plan

- [x] `cd otdf-sdk-mgr && uv run pytest tests/test_schema.py` — all 6 pass
- [x] `uv run python -m otdf_sdk_mgr.schema validate <path>` accepts a valid scenarios.yaml and rejects unknown fields

Jira: https://virtru.atlassian.net/browse/DSPX-3302

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Added schema validation for OpenTDF Scenario and Instance YAML configurations with a new CLI command.
  * Introduced strict validation with cross-field constraints for SDK and platform configurations.

* **Documentation**
  * Updated supported container formats from `nano` to `ztdf-ecwrap`.

* **Dependencies**
  * Updated core package dependencies to support enhanced validation capabilities.

[Review Change Stack](https://app.coderabbit.ai/change-stack/opentdf/tests/pull/450?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

**Stack** (`a60d3302`):

- **#450 DSPX-3302-01-shared-schema ← root** ← this PR
- #451 DSPX-3302-02-platform-installer ← root
- #452 DSPX-3302-03-multi-instance ← root
- #453 DSPX-3302-04-xtest-conftest ↳ on #452
- #454 DSPX-3302-05-claude-plugin ↳ on #453
- #455 DSPX-3302-06-feature-orchestrate ↳ on #454
- #463 DSPX-3382-mlkem-scenarios ↳ on #454
- #485 DSPX-3397-kc26-dpop ↳ on #455

_Generated by [`wgo stack`](https://github.com/dmihalcik-virtru/wgo#stacked-pull-requests). Edit text above or below this block, not inside it._

### CODE REVIEW

File: otdf-sdk-mgr/src/otdf_sdk_mgr/schema.py [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+    args = list(sys.argv[1:] if argv is None else argv)
+    if len(args) != 2 or args[0] != "validate":
+        print("usage: python -m otdf_sdk_mgr.schema validate <path>", file=sys.stderr)
+        return 2
+    path = Path(args[1])
+    try:
+        raw = _yaml().load(path.read_text())
+    except OSError as e:
+        print(f"error: cannot read {path}: {e}", file=sys.stderr)
+        return 1
```
[the remark this answers is not in the document]
Comment (dmihalcik-virtru) [reply otdf-sdk-mgr/src/otdf_sdk_mgr/schema.py:312]: Addressed in 7ff5b63. The validate entry point now reads with explicit UTF-8 and catches YAMLError separately for malformed YAML.

File: otdf-sdk-mgr/src/otdf_sdk_mgr/schema.py [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+    return Instance.model_validate(raw)
+
+
+def dump_instance(instance: Instance, path: str | Path) -> None:
+    """Serialize an Instance to YAML at `path`."""
+    p = Path(path)
+    p.parent.mkdir(parents=True, exist_ok=True)
+    data = instance.model_dump(mode="json", exclude_none=True)
+    y = _yaml()
+    with p.open("w") as f:
```
[the remark this answers is not in the document]
Comment (dmihalcik-virtru) [reply otdf-sdk-mgr/src/otdf_sdk_mgr/schema.py]: Addressed in 7ff5b63. dump_instance now opens the output file with explicit UTF-8 encoding.

File: otdf-sdk-mgr/src/otdf_sdk_mgr/schema.py [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+    elif kind == "Instance":
+        model = Instance
+    else:
+        print(f"error: {path} has unknown kind {kind!r}; expected Scenario or Instance", file=sys.stderr)
+        return 1
+    try:
+        model.model_validate(raw)
+    except Exception as e:
+        print(f"invalid: {e}", file=sys.stderr)
+        return 1
```
[the remark this answers is not in the document]
Comment (dmihalcik-virtru) [reply otdf-sdk-mgr/src/otdf_sdk_mgr/schema.py:335]: Addressed in 7ff5b63. The broad exception handler was narrowed to ValidationError so unexpected programming errors are no longer labeled as invalid input.

File: otdf-sdk-mgr/src/otdf_sdk_mgr/schema.py [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+def _yaml() -> YAML:
+    y = YAML(typ="safe")
+    y.preserve_quotes = True
+    return y
+
+
+def load_scenario(path: str | Path) -> Scenario:
+    """Parse and validate a scenarios.yaml file."""
+    p = Path(path)
+    raw = _yaml().load(p.read_text())
```
[the remark this answers is not in the document]
Comment (dmihalcik-virtru) [reply otdf-sdk-mgr/src/otdf_sdk_mgr/schema.py]: Addressed in 7ff5b63. YAML reads now go through _load_yaml_mapping, which uses read_text with explicit UTF-8 encoding.

File: otdf-sdk-mgr/src/otdf_sdk_mgr/schema.py [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+    instance: Annotated[Instance, Field(description="Inline instance definition")]
+    sdks: ScenarioSdks = Field(default_factory=ScenarioSdks)
+    suite: Suite
+    expected: str | None = None
+    actual: str | None = None
+
+
+def _yaml() -> YAML:
+    y = YAML(typ="safe")
+    y.preserve_quotes = True
```
[the remark this answers is not in the document]
Comment (dmihalcik-virtru) [reply otdf-sdk-mgr/src/otdf_sdk_mgr/schema.py]: Addressed in 7ff5b63. I removed preserve_quotes from the safe YAML loader so the configuration matches the actual loader behavior.

File: otdf-sdk-mgr/src/otdf_sdk_mgr/schema.py [hunk 6]
```diff
@@ -0,0 +1,227 @@
...
+    raw = _yaml().load(p.read_text())
+    if not isinstance(raw, dict):
+        raise ValueError(f"{p}: top-level YAML must be a mapping, got {type(raw).__name__}")
+    return Scenario.model_validate(raw)
+
+
+def load_instance(path: str | Path) -> Instance:
+    """Parse and validate an instance.yaml file."""
+    p = Path(path)
+    raw = _yaml().load(p.read_text())
```
[the remark this answers is not in the document]
Comment (dmihalcik-virtru) [reply otdf-sdk-mgr/src/otdf_sdk_mgr/schema.py]: Addressed in 7ff5b63. YAML reads now use explicit UTF-8 encoding via the shared _load_yaml_mapping helper.

File: otdf-sdk-mgr/src/otdf_sdk_mgr/schema.py [hunk 7]
```diff
@@ -0,0 +1,227 @@
...
+
+from __future__ import annotations
+
+import sys
+from datetime import date
+from pathlib import Path
+from typing import Annotated, Literal
+
+from pydantic import BaseModel, ConfigDict, Field, model_validator
+from ruamel.yaml import YAML
```
[the remark this answers is not in the document]
Comment (dmihalcik-virtru) [reply otdf-sdk-mgr/src/otdf_sdk_mgr/schema.py]: Addressed in 7ff5b63. I imported ValidationError and YAMLError and tightened the validate CLI to handle those specific cases explicitly.

File: otdf-sdk-mgr/src/otdf_sdk_mgr/schema.py [hunk 8]
```diff
@@ -0,0 +1,227 @@
...
+
+API_VERSION = "opentdf.io/v1alpha1"
+
+KasMode = Literal["standard", "key_management"]
+SdkName = Literal["go", "java", "js"]
+ContainerKind = Literal["ztdf", "ztdf-ecwrap", "nano", "nano-with-policy"]
```
[the remark this answers is not in the document]
Comment (dmihalcik-virtru) [reply otdf-sdk-mgr/src/otdf_sdk_mgr/schema.py]: Addressed in 7ff5b63. Suite.containers now uses ContainerKind so allowed container values are validated by the schema.

File: otdf-sdk-mgr/src/otdf_sdk_mgr/schema.py [hunk 9]
```diff
@@ -0,0 +1,227 @@
...
+    sdks: ScenarioSdks = Field(default_factory=ScenarioSdks)
+    suite: Suite
+    expected: str | None = None
+    actual: str | None = None
+
+
+def _yaml() -> YAML:
+    y = YAML(typ="safe")
+    y.preserve_quotes = True
+    return y
```
[the remark this answers is not in the document]
Comment (dmihalcik-virtru) [reply otdf-sdk-mgr/src/otdf_sdk_mgr/schema.py]: Addressed in 7ff5b63. I dropped preserve_quotes from the safe loader so the YAML configuration is not misleading.

File: otdf-sdk-mgr/tests/test_schema.py [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+
+
+def test_unknown_kind_rejected_by_extra_forbid(tmp_path: Path) -> None:
+    bad = tmp_path / "bad.yaml"
+    bad.write_text(
+        "apiVersion: opentdf.io/v1alpha1\nkind: Scenario\nunknown_field: oops\n"
+        "instance:\n  platform: { dist: v0.9.0 }\nsuite:\n  select: foo\n"
+    )
+    with pytest.raises(ValidationError):
+        load_scenario(bad)
```
[the remark this answers is not in the document]
Comment (dmihalcik-virtru) [reply otdf-sdk-mgr/tests/test_schema.py:168]: Addressed in 7ff5b63. I renamed the test to match what it checks and added a separate invalid kind test.

File: otdf-sdk-mgr/src/otdf_sdk_mgr/schema.py [hunk 11]
```diff
@@ -0,0 +1,227 @@
...
+
+
+class Scenario(_StrictModel):
+    """Top-level scenarios.yaml model.
+
+    Composes an Instance with SDK pins and a pytest Suite selection.
+    """
+
+    apiVersion: str = API_VERSION
+    kind: Literal["Scenario"] = "Scenario"
```
[the remark this answers is not in the document]
Comment (dmihalcik-virtru) [reply otdf-sdk-mgr/src/otdf_sdk_mgr/schema.py]: Addressed in 7ff5b63. Scenario.apiVersion and Instance.apiVersion now use Literal["opentdf.io/v1alpha1"], and I added a regression test for an invalid apiVersion.

File: otdf-sdk-mgr/src/otdf_sdk_mgr/schema.py [hunk 12]
```diff
@@ -0,0 +1,227 @@
...
+    image: str | None = None
+
+    @model_validator(mode="after")
+    def _exactly_one(self) -> PlatformPin:
+        set_fields = [k for k in ("dist", "source", "image") if getattr(self, k)]
+        if len(set_fields) != 1:
+            raise ValueError(
+                f"PlatformPin must set exactly one of dist|source|image (got {set_fields or 'none'})"
+            )
+        return self
```
[the remark this answers is not in the document]
Comment (dmihalcik-virtru) [reply otdf-sdk-mgr/src/otdf_sdk_mgr/schema.py:53]: Addressed in 7ff5b63. The exactly-one checks now test field presence with is not None, and there is a regression test covering the empty-string edge case.

File: otdf-sdk-mgr/src/otdf_sdk_mgr/schema.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (dmihalcik-virtru) [reply otdf-sdk-mgr/src/otdf_sdk_mgr/schema.py:335]: Addressed in 7ff5b63. The CLI now catches ValidationError specifically instead of swallowing unrelated exceptions under the invalid label.

File: otdf-sdk-mgr/tests/test_schema.py [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+    assert pin.features["ec_tdf_enabled"] is True
+
+
+def test_scenario_sdks_union_dedupes() -> None:
+    sdks = ScenarioSdks(
+        encrypt={"go": SdkPin(version="lts")},
+        decrypt={"go": SdkPin(version="lts"), "java": SdkPin(version="0.7.8")},
+    )
+    union = sdks.union()
+    assert set(union.keys()) == {"go", "java"}
```
[the remark this answers is not in the document]
Comment (dmihalcik-virtru) [reply otdf-sdk-mgr/tests/test_schema.py]: Addressed in 7ff5b63. The union test now uses conflicting go versions and asserts that the decrypt value wins.

**Review by pflynn-virtru — APPROVED** [review]
